streaming/java/streaming-api/src/main/java/org/ray/streaming/api/stream/Stream.java

```diff
@@ -43,49 +40,67 @@ public Stream(Stream<T> inputStream, StreamOperator streamOperator) {
     this.partition = selectPartition();
   }
 
+  /**
+   * Create a reference of referenced stream.
+   * Changes in new stream will be reflected in referenced stream and vice versa
+   */
+  protected Stream(Stream<T> referencedStream) {
+    this.referencedStream = referencedStream;
```

Comment:
I think we can have a subclass as `ReferencedStream` instead of depending on implicit whether or not `referencedStream == null`. That means, `Stream` as interface and this class renamed as `AbstractStream` and a single implementation of `Stream` named `ReferencedStream` that only take a Stream as field and delegate every operation.

Reply:
Can you elaborate more about `ReferencedStream `? I thought about introducing a new stream class  before. But I can't put it into the class hierarchy. Currently only stream object returned by `asJava/asPython` is reference stream. For example, `DataStream#asPython()` return a `PythonDataStream` which reference previous stream. If we make `DataStream#asPython()` return a `ReferenceStream` instead of `PythonDataStream`, then we can't call `PythonDataStream` related methods.

Reply:
And `PythonDataStream` is not a `ReferencedStream` if it's not a stream returned by `asPython` method. Also, we can't make `PythonDataStream` extends multiple superclasses

Reply:
Well it is more on style which doesn't block feature proposed here. I'll send a follow up PR on this PR when it gets merged. Let's discussion later then.

Reply:
That would be great.